### PR TITLE
Make Spack able to apply gz compressed remote patches

### LINKED ISSUE #22822: Cannot apply a gunzipped patch file
Spack should support applying compressed patches from remote, but currently fails on `.gz` files. The issue seems that `gunzip` always decompress the file in place, while Spack expects it to extract in the current working directory:

https://github.com/spack/spack/blob/58edc361be0aec47bfed24ef310ffa808153c74c/lib/spack/spack/fetch_strategy.py#L465-L472

### Steps to reproduce the issue
Using #22788 modified to retrieve the patch from remote:
```console
$ spack -d patch ncbi-rmblastn
```

### Error Message

<details>

<summary>$ spack -d patch ncbi-rmblastn </summary>

```console
$ spack -d patch ncbi-rmblastn
==> [2021-04-07-10:16:42.722900] Imported patch from built-in commands
==> [2021-04-07-10:16:42.723910] Imported patch from built-in commands
==> [2021-04-07-10:16:42.725527] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2021-04-07-10:16:42.750447] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2021-04-07-10:16:42.821071] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2021-04-07-10:16:42.860430] Reading config file /home/culpo/.spack/packages.yaml
==> [2021-04-07-10:16:42.866125] Reading config file /home/culpo/.spack/linux/compilers.yaml
==> [2021-04-07-10:16:42.971083] Ordered hashes [ncbi-rmblastn]: ncbi-rmblastn/0/ce985abd3512834adb9ad3e4078fbf9608a33a2ee6538a1e94b641490c92f899
==> [2021-04-07-10:16:42.971715] DATABASE LOCK TIMEOUT: 3s
==> [2021-04-07-10:16:42.971787] PACKAGE LOCK TIMEOUT: No timeout
==> [2021-04-07-10:16:42.985895] Creating stage lock spack-stage-ncbi-rmblastn-2.9.0-7v3skwjselu63o5avqqaez3qslpizpr4
==> [2021-04-07-10:16:42.986535] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/mirrors.yaml
==> [2021-04-07-10:16:42.987343] Reading config file /home/culpo/.spack/mirrors.yaml
==> [2021-04-07-10:16:43.073409] Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/a3/a390cc2d7a09422759fc178db84de9def822cbe485916bbb2ec0d215dacdc257.tar.gz
==> [2021-04-07-10:16:43.187731] Creating stage lock spack-stage-g8t8mduw
==> [2021-04-07-10:16:43.188243] Checking existence of http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz
==> [2021-04-07-10:16:43.188815] '/usr/bin/curl' '--stderr' '-' '-s' '-f' '-r' '0-0' 'http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz'
==> [2021-04-07-10:16:43.714717] Fetching http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz
==> [2021-04-07-10:16:43.716420] '/usr/bin/curl' '-C' '-' '-o' '/tmp/culpo/spack-stage/spack-stage-g8t8mduw/isb-2.11.0+-rmblast.patch.gz.part' '-f' '-D' '-' '-L' 'http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz' '-#' '--connect-timeout' '10'
########################################################################################################################################### 100,0%
==> [2021-04-07-10:16:44.404490] Staging archive: /tmp/culpo/spack-stage/spack-stage-g8t8mduw/isb-2.11.0+-rmblast.patch.gz
==> [2021-04-07-10:16:44.407331] '/bin/gunzip' '/tmp/culpo/spack-stage/spack-stage-g8t8mduw/isb-2.11.0+-rmblast.patch.gz'
/tmp/culpo/spack-stage/spack-stage-g8t8mduw/spack-src
==> [2021-04-07-10:16:44.417239] Created stage in /tmp/culpo/spack-stage/spack-stage-g8t8mduw
/tmp/culpo/spack-stage/spack-stage-g8t8mduw/spack-src
==> [2021-04-07-10:16:44.417716] NoSuchPatchError: Archive was empty: http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz
==> [2021-04-07-10:16:44.417856] Error: Archive was empty: http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 772, in main
    return _invoke_command(command, parser, args, unknown)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 496, in _invoke_command
    return_val = command(parser, args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/patch.py", line 36, in patch
    package.do_patch()
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/package.py", line 1397, in do_patch
    self.do_stage()
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/package.py", line 1382, in do_stage
    self.do_fetch(mirror_only)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/package.py", line 1370, in do_fetch
    patch.fetch()
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/patch.py", line 206, in fetch
    "Archive was empty: %s" % self.url)
spack.patch.NoSuchPatchError: Archive was empty: http://www.repeatmasker.org/isb-2.11.0+-rmblast.patch.gz
```

</details>

### Information on your system

* **Spack:** 0.16.1-2091-c188575cfe
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-broadwell
* **Concretizer:** original

### Additional information

The following diff seems to solve the issue:

```diff
diff --git a/lib/spack/spack/util/compression.py b/lib/spack/spack/util/compression.py
index 2f190e94e1..03a830109a 100644
--- a/lib/spack/spack/util/compression.py
+++ b/lib/spack/spack/util/compression.py
@@ -22,6 +22,21 @@ def allowed_archive(path):
     return any(path.endswith(t) for t in ALLOWED_ARCHIVE_TYPES)
 
 
+def _gunzip(archive_file):
+    """Like gunzip, but extracts in cwd instead of in-place.
+
+    Args:
+        archive_file (str): absolute path of the file to be decompressed
+    """
+    import gzip
+    decompressed_file = os.path.basename(archive_file.strip('.gz'))
+    working_dir = os.getcwd()
+    destination_abspath = os.path.join(working_dir, decompressed_file)
+    with gzip.open(archive_file, "rb") as f_in:
+        with open(destination_abspath, "wb") as f_out:
+            f_out.write(f_in.read())
+
+
 def decompressor_for(path, extension=None):
     """Get the appropriate decompressor for a path."""
     if ((extension and re.match(r'\.?zip$', extension)) or
@@ -30,8 +45,7 @@ def decompressor_for(path, extension=None):
         unzip.add_default_arg('-q')
         return unzip
     if extension and re.match(r'gz', extension):
-        gunzip = which('gunzip', required=True)
-        return gunzip
+        return _gunzip
     if extension and re.match(r'bz2', extension):
         bunzip2 = which('bunzip2', required=True)
         return bunzip2

```

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by alalazo**
fixes #22822

When preparing stages etc. we expect the command to decompress an archive to write its output in the current working directory. This is not the case for `gunzip` that instead works in place by default. Here we solve the issue by substituting a system call with the use of the `gzip` module from the Python standard library.

### CODE REVIEW

**Comment by alalazo** [comment]
@glennpj FYI, with this I am able to retrieve patches remotely (in reply to https://github.com/spack/spack/pull/22788#discussion_r607948090)

File: lib/spack/spack/util/compression.py [hunk 1]
```diff
@@ -22,6 +22,22 @@ def allowed_archive(path):
...
 
 
+def _gunzip(archive_file):
+    """Like gunzip, but extracts in the current working directory
+    instead of in-place.
+
+    Args:
+        archive_file (str): absolute path of the file to be decompressed
+    """
+    import gzip
```
Comment (becker33) [inline lib/spack/spack/util/compression.py:32]: Do all python installations support this?

Comment (alalazo) [reply lib/spack/spack/util/compression.py:32]: @becker33 Not sure how to verify this. This module is part of the standard library, and it's present in the Python 2.6 interpreters I have access to. That said I don't know if there are distributions that package it separately like e.g. Ubuntu does for `python-tk`.

An alternative would be to use the old command but try to move the extracted file into place.

**Comment by tgamblin** [comment]
@alalazo what `sha256` ultimately ends up on the spec with gzipped patches?  Is it the patch's SHA (how we do it for uncompressed patches) or the gzip file's SHA?  There are probably not that many places where this *really* matters, but it seems like a consistency issue if the spec says one thing for the gzipped version and another for the non-gzipped version.

**Comment by alalazo** [comment]
@tgamblin Just checked, we record the patch sha, not the gzip sha (so for the hash we'll be transparent on compression of the patch file):

```yaml
spec:
- ncbi-rmblastn:
    version: 2.11.0
    arch:
      platform: linux
      platform_os: ubuntu18.04
      target:
        name: broadwell
        vendor: GenuineIntel
        features:
        - adx
        - aes
        - avx
        - avx2
        - bmi1
        - bmi2
        - f16c
        - fma
        - mmx
        - movbe
        - pclmulqdq
        - popcnt
        - rdrand
        - rdseed
        - sse
        - sse2
        - sse4_1
        - sse4_2
        - ssse3
        generation: 0
        parents:
        - haswell
    compiler:
      name: gcc
      version: 10.3.0
    namespace: builtin
    parameters:
      patches:
      - ce985abd3512834adb9ad3e4078fbf9608a33a2ee6538a1e94b641490c92f899
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    patches:
    - ce985abd3512834adb9ad3e4078fbf9608a33a2ee6538a1e94b641490c92f899
    hash: jnlxw7sgdc7o6fsymcgbgds4vfpztgkb
    full_hash: 2tkyy55sk4k7eooelvsypak2bo5ibly4
    build_hash: jnlxw7sgdc7o6fsymcgbgds4vfpztgkb
```

**Comment by alalazo** [comment]
@glennpj

**Comment by tgamblin** [comment]
@alalazo: 

> Just checked, we record the patch sha, not the gzip sha (so for the hash we'll be transparent on compression of the patch file)

ok and just to be sure, do we also checksum the gz before expanding it? The user should provide the gz SHA, but the patch SHA is what should go on the spec. This is so we don’t expand something we don’t trust (you can embed attacks in compressed formats and it’s very easy for arbitrary code execution to happen).

**Comment by alalazo** [comment]
> ok and just to be sure, do we also checksum the gz before expanding it?

I think this is done here:

https://github.com/spack/spack/blob/46088aa5daa379d6e88859d5de2322b400e6b3e7/lib/spack/spack/patch.py#L225-L236

**Comment by tgamblin** [comment]
ok great -- just checking. This LGTM!

**Review by becker33 — APPROVED** [review]

**Pull request merged**